feat : Redis를 이용한 JWT 로그인 유스케이스 작성(#14) & fix : 지난 코드리뷰 반영

2dongyeop opened the pull request:
## What is this PR? 📍
### *1. GlobalExceptionHandler 도입*
유효성 검증 실패 등의 이유로 예외를 매번 던지다 보면, 개발자의 실수로 인해 예외처리 로직이 누락될 수 있습니다.
따라서 이러한 실수를 방지하기 위해 핸들러를 작성하고, 이곳에서 예외들이 처리되도록 합니다.

### *2. 스프링 시큐리티에 의존하지 않는 비밀번호 암호화 도입*
`JBCrypt`의 경우 Java 기반에서 사용되는 BCrypt이므로 안정적이고 검증되어 널리 사용되는 반면,
`Argon2`는 최신 해시 알고리즘으로, 최신 보안 기준을 준수하는 장점이 있지만 아직은 다른 라이브러리와의 호환성이 보장되지 않는 이유로 JBCrypt를 이용한 비밀번호 암호화를 도입합니다.

### *3. JWT 로그인 유스케이스 작성*
기존에 가입이 된 정보를 바탕으로 로그인 요청을 보내면 유효성 검증을 마친 뒤 `JwtProvider`를 통해 토큰을 발급하도록 합니다.

로그인과 로그아웃의 경우, 1시간을 유효시간으로 설정하여 시간이 지나면 다시 로그아웃을 하도록 제한을 둘 예정입니다.
따라서 이러한 제한 시간을 두기 위해 `Redis`에 토큰을 시간 제한과 함께 저장하도록 하였습니다.

아직 `WebMvcConfig` 등의 접근 제한은 설정하지 않았습니다.
해당 작업을 통해 테스트를 작성하기 번거롭다는 단점으로 인해 기능 개발을 모두 마친 뒤 작성할 예정입니다.


<br/>

## Changes 🔍
### *1. 이메일 유효성 검사 추가*
`@Email`로 이메일 형식을 검사하고, `null`과 공백이 들어오는 것을 방지하기 위해 `@NotBlank`를 함께 사용하였습니다.

### *2. 지난 코드 리뷰 내용들을 반영*
- 엔드포인트 매핑시 중복된 URL을 `@RequestMapping`으로 분리하였습니다. - 주영님 리뷰
- 매번 엔티티마다 `NotFoundException`을 만들어 예외를 처리하다보니 커스텀 예외가 너무 많아져 이를 수정하였습니다. - 현아님 리뷰

<br/>
 
## Todo 🗓️
- [ ] 로그아웃 기능 구현하기
- [ ] 이메일 인증 기능 추가하기
- [ ] 기능 구현이 완료되면, WebMvcConfig 작성하기


<br/>

2dongyeop committed:
fix : 이메일 유효성 검사 추가

웹 클라이언트로부터 받는 SignupRequest에서 받는 email 유효성 검사를 위한 애너테이션을 추가합니다.
`@Email` 과 함께 `@NotBlank`를 사용한 이유는 해당 어노테이션은 null과 ""(공백)을 유효하다고 판단하기 때문에
이를 방지하기 위해 `@NotBlank`를 함께 사용하였습니다.

참고하면 좋은 자료입니다.
- https://bellog.tistory.com/134

2dongyeop committed:
feat : BoardMapper를 의존성 주입이 아닌 static 메서드 호출 방식으로 변환

2dongyeop committed:
fix : 중복된 URI를 으로 분리

2dongyeop committed:
feat : Exception handler 생성 및 적용

여러 유효성 검사마다 예외를 던지고, 해당 메소드 혹은 호출한 메소드에서 예외를 매번 처리해주기에는 무리가 있을 수 있습니다. (개발자의 실수로 예외처리 누락 등..)

따라서 Global Exception Handler를 만들어 예외를 관리하도록 합니다.

2dongyeop committed:
feat : 스프링 시큐리티를 사용하지 않은 비밀번호 암호(#14)

회원가입 및 의료진가입시 요청으로 들어온 비밀번호를 암호화 합니다.
암호화 라이브러리에는 Spring Security에서 기본적으로 제공하는 BCrypt를 포함해, Java 구현체인 jBCrypt와 PBKDF2, Argon2 등이 있습니다.

아직 스프링 시큐리티를 잘 알지 못하기에 의존하지 않으려 jBCrypt와 Argon2 중에 고민하였고,
jBCrypt의 경우 Java 기반에서 사용되는 BCrypt이므로 안정적이고 검증되어 널리 사용되는 반면
Argon2는 최신 해시 알고리즘으로, 최신 보안 기준을 준수하는 장점이 있지만 아직은 다른 라이브러리와의 호환성이 보장되지 않습니다.

이러한 이유로 jBCrypt를 이용한 비밀번호 암호화를 이용하기로 결정하였습니다.

Related to : #14

2dongyeop committed:
feat : 회원 가입시 비밀번호와 함께 확인 비밀번호를 요청하도록 유효성 검사 추가

회원 가입 및 의료진 가입의 경우, 비밀번호를 한번만 입력받을 경우 사용자가 잘못된 비밀번호를 입력할 수도 있습니다.
따라서 확인 비밀번호를 통해 다시 한번 비밀번호를 입력받아, 실수를 피할 수 있도록 합니다.

2dongyeop committed:
feat : 스프링 시큐리티를 사용하지 않은 비밀번호 암호(#14)

회원가입 및 의료진가입시 요청으로 들어온 비밀번호를 암호화 합니다.
암호화 라이브러리에는 Spring Security에서 기본적으로 제공하는 BCrypt를 포함해, Java 구현체인 jBCrypt와 PBKDF2, Argon2 등이 있습니다.

아직 스프링 시큐리티를 잘 알지 못하기에 의존하지 않으려 jBCrypt와 Argon2 중에 고민하였고,
jBCrypt의 경우 Java 기반에서 사용되는 BCrypt이므로 안정적이고 검증되어 널리 사용되는 반면
Argon2는 최신 해시 알고리즘으로, 최신 보안 기준을 준수하는 장점이 있지만 아직은 다른 라이브러리와의 호환성이 보장되지 않습니다.

이러한 이유로 jBCrypt를 이용한 비밀번호 암호화를 이용하기로 결정하였습니다.

Related to : #14

2dongyeop committed:
feat : 스프링 시큐리티를 사용하지 않은 비밀번호 암호(#14)

회원가입 및 의료진가입시 요청으로 들어온 비밀번호를 암호화 합니다.
암호화 라이브러리에는 Spring Security에서 기본적으로 제공하는 BCrypt를 포함해, Java 구현체인 jBCrypt와 PBKDF2, Argon2 등이 있습니다.

아직 스프링 시큐리티를 잘 알지 못하기에 의존하지 않으려 jBCrypt와 Argon2 중에 고민하였고,
jBCrypt의 경우 Java 기반에서 사용되는 BCrypt이므로 안정적이고 검증되어 널리 사용되는 반면
Argon2는 최신 해시 알고리즘으로, 최신 보안 기준을 준수하는 장점이 있지만 아직은 다른 라이브러리와의 호환성이 보장되지 않습니다.

이러한 이유로 jBCrypt를 이용한 비밀번호 암호화를 이용하기로 결정하였습니다.

Related to : #14

2dongyeop committed:
feat : JWT 로그인 유스케이스 작성, Redis 도입(#14)

로그인 유스케이스를 작성합니다.
기존에 가입이 된 정보를 바탕으로 로그인 요청을 보내면 유효성 검증을 마친 뒤
JwtProvider를 통해 토큰을 발급하도록 합니다.

로그인과 로그아웃의 경우, 1시간을 유효시간으로 설정하여 시간이 지나면 다시 로그아웃을 하도록 제한을 둘 예정입니다.
이를 위해 토큰을 Redis에 시간 제한과 함께 저장합니다.

아직 WebMvcConfig 등의 접근 제한은 설정하지 않았습니다.
해당 작업을 통해 테스트를 작성하기 번거롭다는 단점으로 인해 기능 개발을 모두 마친 뒤 작성할 예정입니다.

Related to : #14

2dongyeop commented on src/main/java/io/wisoft/avocadobackendhexagonal/domain/auth/adapter/in/web/dto/LoginRequest.java:
```diff
@@ -0,0 +1,9 @@
+import jakarta.validation.constraints.NotBlank;
+
+public record LoginRequest(
+        @NotBlank String email,
```
`@Email` 어노테이션이 빠져있네요. 추가하겠습니다..~

bigtyno931128 commented on src/main/java/io/wisoft/avocadobackendhexagonal/domain/auth/adapter/in/web/SignupController.java:
안녕하세요 . 주영님 개인적인 의견입니다 ! 보통 비밀번호 와 확인 비밀번호 를 입력하여 회원가입할 시 
확인 비밀번호의 경우에는 화면에서 바로 패턴체크를 하고 비밀번호 만 넘기는 방식을 많이 사용 하는것 같아서 서버까지 확인 비밀번호를 넘겨서 체크하는 이유가 있을까요 ?

2dongyeop replied:
제가 의도한 바는 "영문, 숫자 등이 섞이고.. 길이는 8자 이상.." 등의 정규식 패턴이 아닌, 
[첨부 링크와 같이](https://www.google.com/url?sa=i&url=https%3A%2F%2Fwww.inflearn.com%2Fquestions%2F240928%2F%25EB%25B9%2584%25EB%25B0%2580%25EB%25B2%2588%25ED%2598%25B8-%25EC%259D%25BC%25EC%25B9%2598-%25ED%2599%2595%25EC%259D%25B8-%25EC%25BD%2594%25EB%2593%259C&psig=AOvVaw3SkZ7rYPRGn64ipi3BNTXK&ust=1686667256011000&source=images&cd=vfe&ved=0CBEQjRxqFwoTCKDHran7vf8CFQAAAAAdAAAAABAN) `비밀번호` 와 `비밀번호 확인` 폼을 만들어 두 값이 일치하지 않으면 다시 입력하도록 하는 작업을 의도하였습니다.

개인적인 생각으로는 프론트에서 처리한다고 해도, 백엔드에서 동일하게 애플리케이션을 위해 계층별로 적합한 유효성 검사를 진행해야 한다고 생각합니다.

**별개로.. 비밀번호 정규식을 검사하는 로직을 추가하기는 해야겠네요. 다음 PR때 함께 수정하겠습니다.**

bigtyno931128 replied:
이해했습니다 ! 자료도 같이 답변주시니 이해가 편한것 같습니다 . 핵사고날 아키텍쳐에 대해서 잘 모르다보니 리뷰가 적었던것 같습니다 . 저도 이번 기회에 학습하여 더 도움 되는 리뷰 남기도록 하겠습니다 .

Yana94Ko replied:
백엔드까지 가지고 오는건 처음봐서 굉장히.. 새롭네요

bigtyno931128 commented on src/main/java/io/wisoft/avocadobackendhexagonal/domain/auth/application/service/LoginService.java:
추후에 업데이트 된다면 access token refresh token 로 구현해보는것도 좋을것 같습니다 !

2dongyeop replied:
맞네요! 현재 엑세스 토큰만을 다루고 있으니.. 다음 PR에서 반영하겠습니다!

bigtyno931128 commented on src/main/java/io/wisoft/avocadobackendhexagonal/domain/board/adapter/in/web/DeleteBoardController.java:
```diff
@@ -5,15 +5,17 @@
 
-    @DeleteMapping("/api/boards/{id}")
+    @DeleteMapping("/{id}")
     public ResponseEntity<Void> deleteBoard(@PathVariable final Long id) {
```
id 를 final 로 받은 이유가 궁금합니다 .

2dongyeop replied:
자바로 함수형 프로그래밍을 구현할 경우, 함수나 메서드는 지역변수만을 변경해야 함수형이라 할 수 있는데,
이를 위해서는 함수나 메서드에서 참조하는 객체(예를 들면 파라미터로 넘어온 값)는 **불변**객체여야 합니다.

다른 이유로는 **"Thread-Safe하여 병렬 프로그래밍에 유용하며, 동기화를 고려하지 않아도 된다"** 와 
**"부수 효과(Side Effect)를 피해 오류가능성을 최소화할 수 있다"** 가 있습니다.


참고하면 좋은 자료입니다.
- https://mangkyu.tistory.com/131
- https://xrabcde.github.io/modernjava-ch18/
- 모던 자바 인 액션 & 클린 코드

bigtyno931128 replied:
와 많이 배우고 갑니다 . 감사합니다 .

Yana94Ko replied:
deleteBoard라는 메서드 내부에서만 사용되는 id 값인데, 멀티스레드와 관련이 있을까요?
하나의 메서드를 여러개의 스레드가 가져다가 쓰지 않을텐데 Thread-Safe와 메서드 매개변수의 final과 연관이 있을지 궁금합니다!

2dongyeop replied:
@Yana94Ko 아래 설명("다른 이유로는~" 부터)에서는 현재 코드가 아닌 `final` 키워드 사용의 이점에 대한 부가 설명이였습니다.
현재 상황에서는 **(setter를 만들지 않는 것과 같이) 불변 객체임을 지킴으로써 부수 효과를 방지하기 위함**이라고 보시면 됩니다!

2dongyeop committed:
fix : 지난 코드 리뷰 반영 - 비밀번호 정규 패턴 검사 로직 추가 및  검사 추가

Yana94Ko commented on src/main/java/io/wisoft/avocadobackendhexagonal/domain/auth/adapter/in/web/LoginController.java:
```diff
@@ -0,0 +1,43 @@
+
+    private final LoginUseCase loginUseCase;
+
+    @PostMapping("/login/members")
+    public ResponseEntity<TokenResponse> signupMember(@RequestBody @Valid final LoginRequest request) {
+
+        final LoginCommand loginCommand = getLoginCommand(request);
+        final String token = loginUseCase.loginMember(loginCommand);
+        return ResponseEntity.ok(new TokenResponse(token, "bearer"));
+    }
+
+
+    @PostMapping("/login/staff")
+    public ResponseEntity<TokenResponse> loginStaff(@RequestBody @Valid final LoginRequest request) {
+
+        final LoginCommand loginCommand = getLoginCommand(request);
+        final String token = loginUseCase.loginStaff(loginCommand);
+        return ResponseEntity.ok(new TokenResponse(token, "bearer"));
+    }
+
+    private LoginCommand getLoginCommand(final LoginRequest request) {
+        final LoginCommand loginCommand = new LoginCommand(request.email(), request.password());
+        return loginCommand;
+    }
```
[현업] getLoginCommand() 메소드를 정의해서 가져다 쓰기보다는, LoginCommand 클래스에 LoginRequest 객체를 받아서 LoginCommand 객체를 반환하는 메소드를 만들어서 쓰는게 어떨까요?
```
final LoginCommand command = LoginCommand.from(loginRequest)
```

2dongyeop replied:
@Yana94Ko  우선 리뷰해주시느라 고생하셨습니다!
현재 상황에서 `Request`는 인커밍 어댑터에서 쓰이는 dto이고 `Command`는 인커밍 포트에서 쓰이는 dto입니다.

말씀하신대로  해당 생성자를 만든다면 `Command` 내부에 `Request`가 들어가게 되는데 (첨부 그림에서 초록색 화살표), 
이는 클린아키텍처에서 말하는 ***바깥으로 나가는 의존성은 없어야 한다***(그림에서 주황색 화살표)에 어긋나지 않을까 생각이 듭니다..

![스크린샷 2023-06-14 오후 10 24 15](https://github.com/HBNU-Avocado/Avocado-backend-with-hexagonal/assets/95692663/acca624b-500a-4be2-b2bc-744357f9dae4)

Yana94Ko replied:
**생성자를 만드는 것은 아니고**, `LoginCommand`에 `LoginRequest`를 인수로 받아 `LoginCommand` 객체를 반환하는 **정적 메소드를 만드는것입니다**. 이는 정적 메소드로 정의하는 것이기 때문에 `LoginCommand` 객체를 만드는데는 영향을 미치지 않게 되므로, 의존성이 생긴다고 보기는 어렵지 않을까요?

2dongyeop replied:
@Yana94Ko  음.. 의존 관계와 연관 관계를 생각해서 계층별로 의존도를 따지면 아래 그림처럼 되는데,
헥사고날 아키텍처의 가장 큰 특징에는 의존성 역전 원칙이 있고, 쉽게 말하면 **내부 계층은 바깥 계층을 알지 못하는 것**입니다.

계층형 아키텍처로 예를 들면 "**http와 같은 프로토콜에 관한 로직은 controller에서 담당**"하고,
"**service는 비즈니스 로직만을 담당하여 controller 계층에서 어떤 프로토콜을 사용하는지는 알면 안된다.**"와 비슷합니다.

지금 이 계층 간 dto 변환의 경우에서는, 웹 어댑터(= 인커밍 어댑터)인 컨트롤러에서 어댑터의 역할로써 변환해주는게 맞다는 주관적인 생각이 드네요! 클린 아키텍처 공부 중에 [8장 초반에 나온 "각 계층에 대해 모델을 만드는 것은 항상 논란거리이다."](https://leedongyeop.notion.site/08-7fe912839d6a4e5da228ac8e5a21968b?pvs=4)가 이 내용인가 봅니다 😂


별개로 아래는 의존성에 대해서 공부할 때 참고한 자료들인데, 좋은 자료였어서 공유드립니다!
- [우아한 객체지향](https://javanitto.tistory.com/41)
- [우아한 객체지향 youtube](https://www.youtube.com/watch?v=dJ5C4qRqAgA)

<img width="732" alt="스크린샷 2023-06-15 오전 12 20 38" src="https://github.com/HBNU-Avocado/Avocado-backend-with-hexagonal/assets/95692663/8faf3cd4-4e71-4980-b36a-72eb186b3083">

Yana94Ko replied:
![image](https://github.com/HBNU-Avocado/Avocado-backend-with-hexagonal/assets/91898700/7b551f22-6ed1-45ee-aaad-2af49217f5f9)

모각코가 끝나서 리뷰댓글 캡쳐로 전달..! 드립니다..!

Yana94Ko commented on src/main/java/io/wisoft/avocadobackendhexagonal/domain/hospital/adapter/in/web/LoadHospitalController.java:
```diff
@@ -1,48 +1,48 @@
+    }
+
     private HospitalDto getHospitalDto(final Hospital hospital) {
-        final HospitalDto hospitalDto = new HospitalDto(
+        return new HospitalDto(
                 hospital.getId(),
                 hospital.getName(),
                 hospital.getNumber(),
                 hospital.getAddress(),
                 hospital.getOperatingTime()
         );
-        return hospitalDto;
-    }
```
해당 DTO의 경우에도 꼭 컨트롤러에서 함수를 생성했어야 했는지 생각해보면 좋을것 같아요!

2dongyeop replied:
이 경우에는 별로 중요하지 않은 DTO에서 중요한 Domain에 의존하는 것이니, 컨트롤러에서 작성하는게 오히려 코드를 더럽힐 수 있겠네요!
다음 PR때 반영해서 수정하겠습니다. 👍

Yana94Ko commented on build.gradle:
```diff
@@ -22,6 +22,10 @@ dependencies {
 	implementation 'org.springframework.boot:spring-boot-starter-validation'
+
+	// https://mvnrepository.com/artifact/org.mindrot/jbcrypt
+	implementation group: 'org.mindrot', name: 'jbcrypt', version: '0.4'
```
jBCrypt 얻어갑니다.. 감사합니동 희희
node에서 user 구현시에 bcrypt썼었는데, 자바에서도 사용 가능하다니 당장 써봐야겠네요!

2dongyeop replied:
시큐리티 자체로도 공부할 내용이 많아 부담스러우니 면접때 대답 못 할 바에는.. 이게 나은 것 같네요!

Yana94Ko commented on build.gradle:
```diff
@@ -22,10 +22,6 @@ dependencies {
-
-	// https://mvnrepository.com/artifact/org.mindrot/jbcrypt
-	implementation group: 'org.mindrot', name: 'jbcrypt', version: '0.4'
-
```
얘는 어디로 사라진걸까요..! 다시 bcrypt로 넘어가시는걸까요!

2dongyeop replied:
커밋 과정에서 revert/reset 실수로 번복 커밋이 된 것 같네요. 주의하겠습니다 😂

Yana94Ko commented on src/main/java/io/wisoft/avocadobackendhexagonal/domain/auth/application/service/SignUpService.java:
```diff
@@ -24,6 +25,7 @@ public class SignUpService implements SignupUseCase {
     private final SaveStaffPort saveStaffPort;
     private final LoadStaffPort loadStaffPort;
     private final LoadHospitalPort loadHospitalPort;
+    private final EncryptHelper encryptHelper;
```
동일한 commit message 로 커밋이 두번 일어났던데
두번째 커밋에서는 어떤 변화가 있던걸까요?
feat 인지 fix인지 commit convention을 확인 부탁드려요!

2dongyeop replied:
커밋 과정에서 revert/reset 실수로 번복 커밋이 된 것 같네요. 주의하겠습니다 😂

2dongyeop committed:
feat : Redis 관련 코드의 가독성 및 사용성을 증진시키기 위한 Redis Adapter 생성 및 적용

Redis 관련 crud 코드들은 `ValueOperations<String, String>` 등과 같은 키워드들로 인해 가독성이 떨어지기 때문에, 메서드를 매핑함으로써 사용성도 챙기도록 하였습니다.

JWT가 적재되지 않았을 경우에 대한 예외처리가 없어서 추가로 작성하였습니다.

2dongyeop committed:
feat : Refresh token을 이용해 Access token 재발급받기(#16)

보안을 위해 유효시간이 비교적 짧게 설정된 access token이 만료될 경우, 사용자는 로그인시 access token과 함께 발급받은 refresh token을 이용해 access token을 재발급 받아야 합니다.

현재 로그인시 생성된 리프레시 토큰을 redis에 키(이메일)-값(리프레시 토큰)의 형태로 저장하고 있습니다.

리프레시 토큰을 헤더에 싣고, 재발급 요청이 들어오면 아래 과정을 거친 후 액세스 토큰을 재발급합니다.
1. 헤더로부터 bearer 토큰(=JWT) 추출
2. JWT로부터 subject(=email) 추출
3. Redis에 email과 일치하는 키-값이 존재하는지 확인
4. 존재하면 액세스 토큰 재발급

Related to : #16

2dongyeop committed:
refactor : 도메인을 dto로 변환하는 책임을 컨트롤러가 아닌 dto로 이전

Yana94Ko commented on src/main/resources/application.yml:
```diff
@@ -1,6 +1,6 @@
 spring:
   profiles:
-    include: db
+    include: db, jwt, redis
```
`jwt`, `redis` 프로파일은 단순히 api 키를 숨기기 위해서 별도로 생성하신걸까요?

2dongyeop replied:
네 모두 비밀번호나 인코딩 시크릿키 등이 존재하여 관련 분야별로 나누고, `gitignore` 해놓았습니다.

Yana94Ko replied:
아 다른 확장기능을 설정하시는데에 쓰셨는지 궁금해하셔서 관련 코멘트 남겼었어요!
